uwblueprint/paramedics-web
# Add linting

## Overview
>[Code review doc for reference](https://www.notion.so/uwblueprintexecs/Code-Review-a21fc85b00394f488e92d9d605f6b2bc)

[Notion task](https://www.notion.so/uwblueprintexecs/docker-linter-setup-e63c6fbc895b4f4e904e228bab13582b)

**SEE MY INLINE COMMENTS FOR MOST RELEVANT CHANGES**

## Changes
- linted all files based on linter set up in this commit: https://github.com/uwblueprint/paramedics-web/commit/6c05f55f53b0c6d9042e1d5ba5f46aea1c5e486f
  - refactored some code for linting and performance
- reverted `eventAmbulances` and `eventHospitals` tables to be non-paranoid
  - refactored `addAmbulancesToEvent` and `addHospitalsToEvent` as they were really unwieldy with soft deletes
- set up `dev-lint` job in circleCI
  - refactored existing jobs/steps for performance
    - `dev-build` builds using dev dockerfile
    - `prod-build` builds using prod dockerfile (new)
    - `dev-lint` lints using dev dockerfile as that's where we installed eslint (new)
    - removed `install` step as it wasn't doing anything

## Testing
- running `npm run lint` locally (or via vscode) has no errors
- verify that existing resolvers are working as expected (especially `event`)
- test out `addAmbulancesToEvent` and `addHospitalsToEvent`
  - add and delete resources and check the db tables to make sure all changes look correct
- look with your eyes at the checks on this PR and see that all jobs have a green check mark

## Notes
- we should move towards preferring promise chaining over `await` statements - the latter can still be used when necessary
- in general, our CI/CD pipeline is really small and thus fairly fast - we can break down the steps in our `docker-compose` file and optimize even more for performance, but it's not that expensive (i.e. slow) to just do everything from scratch with each build
- not entirely convinced that the caching within jobs are even doing anything - removed for now
  - this is related to the fact that we simply use `docker-compose up` to build the container, which would not get the benefit of caching `package-lock.json` - we could optimize by breaking it down into discrete steps ([more info here](https://circleci.com/docs/2.0/caching/#section=projects))
